drivers/flash/flash_andes_qspi.c

```diff
@@ -0,0 +1,950 @@
+{
+	int ret;
+	uint8_t reg;
+
+	do {
+		ret = flash_andes_qspi_cmd_read(dev,
+		FLASH_ANDES_CMD_RDSR, &reg, 1);
+	} while (!ret && (reg & FLASH_ANDES_WIP_BIT));
+
+	return ret;
+}
+
+#if defined(CONFIG_FLASH_ANDES_QSPI_SFDP_RUNTIME)	\
+	|| defined(CONFIG_FLASH_JESD216_API)
+/*
+ * @brief Read content from the SFDP hierarchy
+ *
+ * @note The device must be externally acquired before invoking this
+ * function.
+ *
+ * @param dev Device struct
+ * @param addr The address to send
+ * @param data The buffer to store or read the value
+ * @param length The size of the buffer
+ * @return 0 on success, negative errno code otherwise
+ */
+static int read_sfdp(const struct device *const dev,
+		     off_t addr, void *data, size_t length)
+{
+	/* READ_SFDP requires a 24-bit address followed by a single
+	 * byte for a wait state.  This is effected by using 32-bit
+	 * address by shifting the 24-bit address up 8 bits.
+	 */
+	return flash_andes_qspi_access(dev, JESD216_CMD_READ_SFDP,
+			ANDES_ACCESS_ADDRESSED,
+			addr, data, length);
+}
+#endif /* CONFIG_FLASH_ANDES_QSPI_SFDP_RUNTIME */
+
+
+/**
+ * @brief Write the status register.
+ *
+ * @note The device must be externally acquired before invoking this
+ * function.
+ *
+ * @param dev Device struct
+ * @param sr The new value of the status register
+ *
+ * @return 0 on success or a negative error code.
+ */
+static int flash_andes_qspi_wrsr(const struct device *dev,
+			uint8_t sr)
+{
+	int ret = flash_andes_qspi_cmd_write(dev, FLASH_ANDES_CMD_WREN);
+
+	if (ret == 0) {
+		ret = flash_andes_qspi_access(dev, FLASH_ANDES_CMD_WRSR,
+					ANDES_ACCESS_WRITE, 0, &sr,
+					sizeof(sr));
+		flash_andes_qspi_wait_until_ready(dev);
+	}
+
+	return ret;
+}
+
+static int flash_andes_qspi_read(const struct device *dev,
+		off_t addr, void *dest, size_t size)
+{
+	const size_t flash_size = dev_flash_size(dev);
+	int ret;
+
+	/* should be between 0 and flash size */
+	if ((addr < 0) || ((addr + size) > flash_size)) {
+		return -EINVAL;
+	}
+
+	acquire_device(dev);
+
+	ret = flash_andes_qspi_cmd_addr_read(dev,
+		FLASH_ANDES_CMD_4READ, addr, dest, size);
+
+	release_device(dev);
+	return ret;
+}
+
+static int flash_andes_qspi_write(const struct device *dev, off_t addr,
+			 const void *src,
+			 size_t size)
+{
+	const size_t flash_size = dev_flash_size(dev);
+	const uint16_t page_size = dev_page_size(dev);
+	int ret = 0;
+
+	/* should be between 0 and flash size */
+	if ((addr < 0) || ((size + addr) > flash_size)) {
+		return -EINVAL;
+	}
+
+	acquire_device(dev);
+
+	ret = flash_andes_qspi_write_protection_set(dev, false);
+
+	if (ret == 0) {
+		while (size > 0) {
+			size_t to_write = size;
+
+			/* Don't write more than a page. */
+			if (to_write >= page_size) {
+				to_write = page_size;
+			}
+
+			/* Don't write across a page boundary */
+			if (((addr + to_write - 1U) / page_size) !=
+			(addr / page_size)) {
+				to_write = page_size - (addr % page_size);
+			}
```

Comment:
This adjustment is required only for the write of the first chunk, as once you aligne the addr you can write by pages. This means that this code should be outside of loop.
Only check and `to_write` adjustment should is needed for final chunk where `size`  may be smaller than page size.

Reply:
Right, boundary checking is redundant after sending first chunk.
As a result, I modified the code to get the size in line 356, so that I don't have to check boundary and size of `to_write`.